### ISSUE TITLE
xrootd: Http and Macarrons deployment

### ISSUE THREAD

**Issue by cabrillo** [xrootd/xrootd#1460]
Dear,
  I nee help to deploy http plugin for xrootd servers.  The xrootd version we are runnin right now is xrootd-server-5.1.1-1.
  For sec.protocol we are using xrootd-voms-pluging and works fine.

But with the http/macaroons pluging we have some troubles, that is the config file:
if exec xrootd
  xrd.protocol http:1094 /usr/lib64/libXrdHttp.so
  http.secxtractor libXrdHttpVOMS.so
  http.cadir /etc/grid-security/certificates/
  http.cert /etc/grid-security/xrd/xrdcert.pem
  http.key /etc/grid-security/xrd/xrdkey.pem
  http.exthandler xrdtpc libXrdHttpTPC.so
  http.exthandler xrdmacaroons libXrdMacaroons.so
  http.desthttps yes
  http.trace all
  macaroons.secretkey /etc/xrootd/macaroon-secret
  #macaroons.onmissing allow
  macaroons.trace all
  ofs.authlib libXrdMacaroons.so
  http.listingdeny yes
  http.staticpreload http://static/robots.txt /etc/xrootd/robots.txt  
  http.header2cgi Authorization authz
fi

With this conf, i always get the same result using my cert:
(Davix::HttpRequest) Error: HTTP 403 : Permission refused 
my user seems to be mapped to "**locsices**" user but the http group seems not to be enought :
210520 14:34:49 21556 anon.0:76@login1 sysXrdHttp:  Extracting auth info.
http Protocol 'gsi'
http Name 'locsices'
http Host '[::ffff:10.10.150.1]'
http Vorg 'cms cms'
http Role 'NULL NULL'
http Grps '/cms /cms/escms'

authfile allow the group /escms to write:
t writecmsdata  /store/temp/              a  \
g /escms          readcmsdata writecmsdata

210520 14:48:34 21178 locsices.320:43@login1 XrootdResponse: sending err 3010: Unable to create /store/temp/stor.davix.06
[davix-put.log](https://github.com/xrootd/xrootd/files/6516110/davix-put.log)
[xrootd-clustered.log](https://github.com/xrootd/xrootd/files/6516114/xrootd-clustered.log)
; permission denied

and macaroon-init doesn't work too:

macaroon-init https://pool09.ifca.es//store 60 DOWNLOAD,UPLOAD
Failed to retrieve macaroon: Macaroon request failed: Could not connect to server
(if   #macaroons.onmissing allow is enable we can read/write without any restriction even without a valid proxy)

complete xrootd conf file and error log output are attached.

Any advise is welcomed, Regards, I

**Comment by bbockelm** [comment]
Looks like you've authorized the group `/escms`:

```
g /escms readcmsdata writecmsdata
```

But use a VOMS FQAN with group `/cms/escms`:
```
http Grps '/cms /cms/escms'
```

Perhaps try changing the authfile to:
```
g /cms readcmsdata writecmsdata
```

For the macaroons:
```
macaroon-init https://pool09.ifca.es//store 60 DOWNLOAD,UPLOAD
Failed to retrieve macaroon: Macaroon request failed: Could not connect to server
```

It appears that you are using the default port above in the client (443) but have configured XRootD to listen on 1094:
```
xrd.protocol http:1094 /usr/lib64/libXrdHttp.so
```

**Comment by xrootd-dev** [comment]
I think that Brian's suggestion is correct but that is not the root cause 
of the problem. Here the problem appears to be:

Failed to retrieve macaroon: Macaroon request failed: Could not connect to 
server

So, that means that the macaroon could not be validated because the issuer 
was simply not available. BTW none of the logs supplied include this 
message so the context is missing. Perhaps, you can include that log as 
well.

Andy


On Thu, 20 May 2021, cabrillo wrote:

> Dear,
>  I nee help to deploy http plugin for xrootd servers.  The xrootd version we are runnin right now is xrootd-server-5.1.1-1.
>  For sec.protocol we are using xrootd-voms-pluging and works fine.
>
> But with the http/macaroons pluging we have some troubles, that is the config file:
> if exec xrootd
>  xrd.protocol http:1094 /usr/lib64/libXrdHttp.so
>  http.secxtractor libXrdHttpVOMS.so
>  http.cadir /etc/grid-security/certificates/
>  http.cert /etc/grid-security/xrd/xrdcert.pem
>  http.key /etc/grid-security/xrd/xrdkey.pem
>  http.exthandler xrdtpc libXrdHttpTPC.so
>  http.exthandler xrdmacaroons libXrdMacaroons.so
>  http.desthttps yes
>  http.trace all
>  macaroons.secretkey /etc/xrootd/macaroon-secret
>  #macaroons.onmissing allow
>  macaroons.trace all
>  ofs.authlib libXrdMacaroons.so
>  http.listingdeny yes
>  http.staticpreload http://static/robots.txt /etc/xrootd/robots.txt
>  http.header2cgi Authorization authz
> fi
>
> With this conf, i always get the same result using my cert:
> (Davix::HttpRequest) Error: HTTP 403 : Permission refused
> my user seems to be mapped to "**locsices**" user but the http group seems not to be enought :
> 210520 14:34:49 21556 ***@***.*** sysXrdHttp:  Extracting auth info.
> http Protocol 'gsi'
> http Name 'locsices'
> http Host '[::ffff:10.10.150.1]'
> http Vorg 'cms cms'
> http Role 'NULL NULL'
> http Grps '/cms /cms/escms'
>
> authfile allow the group /escms to write:
> t writecmsdata  /store/temp/              a  \
> g /escms          readcmsdata writecmsdata
>
> 210520 14:48:34 21178 ***@***.*** XrootdResponse: sending err 3010: Unable to create /store/temp/stor.davix.06
> [davix-put.log](https://github.com/xrootd/xrootd/files/6516110/davix-put.log)
> [xrootd-clustered.log](https://github.com/xrootd/xrootd/files/6516114/xrootd-clustered.log)
> ; permission denied
>
> and macaroon-init doesn't work too:
>
> macaroon-init https://pool09.ifca.es//store 60 DOWNLOAD,UPLOAD
> Failed to retrieve macaroon: Macaroon request failed: Could not connect to server
> (if   #macaroons.onmissing allow is enable we can read/write without 
any restriction even without a valid proxy)

**Comment by cabrillo** [comment]
Dear Andy,
   There was a port missing (no port 443) while i had configured the 1094 (that was fixed with no luck).
   But the autentication issue still continue, for http (write), davx can list and read but not write, and macaroon cannot do anything.
   (log attached at the before post)
regards, I

> I think that Brian's suggestion is correct but that is not the root cause of the problem. Here the problem appears to be: Failed to retrieve macaroon: Macaroon request failed: Could not connect to server So, that means that the macaroon could not be validated because the issuer was simply not available. BTW none of the logs supplied include this message so the context is missing. Perhaps, you can include that log as well. Andy
>

**Comment by cabrillo** [comment]
HI Guys,

   Any other advise, about this config, I have been trying other configurations with no luck, seems that libXrdHttpVOMS and authfile are not working fine together at least in v 5.1.1-1 

Regards, I

**Comment by wyang007** [comment]
what is "readcmsdata" ?

> Hi Brian,
> That's the actual Authfile conf:
> 
> t writecmsuser /store/temp/ a 
> /store/user/ a
> 
> g /cms readcmsdata writecmsuser
> g /cms/escms readcmsdata writecmsuser
> g /cms /cms/escms readcmsdata writecmsuser# in the case the http grp require whole string
> g /escms readcmsdata writecmsuser

**Comment by cabrillo** [comment]
[root@pool09 ~]# cat /etc/xrootd/Authfile 
t writecmsdata /store/backfill/          a  \
               /store/data/              a  \
               /store/generator/         a  \
               /store/group/             a  \
               /store/hidata/            a  \
               /store/himc/              a  \
               /store/mc/                a  \
               /store/PhEDEx_LoadTest07/ a  \
               /store/relval/            a  \
               /store/results/           a  \
               /store/temp/              a  \
               /store/unmerged/          a  \
               /store/user/              a

t readcmsdata  /store/unmerged/          a  \
                        /store/                   lr

t writecmsuser /store/temp/              a  \
                        /store/user/              a

# CMS users have full access to their own directory, and read for CMS
# While xrootd allows the user to *attempt* any operation - even in other user's
# home directories - the underlying filesystem also has its internal permissions and will further
# limit things.

g /cms            readcmsdata writecmsuser
g /escms          readcmsdata writecmsdata
g /cms/GGUSExpert readcmsdata writecmsdata
g /cms/TEAM       readcmsdata writecmsdata
g /cms/ALARM      readcmsdata writecmsdata

this works fine for xrootd but not for http I have tried to use (o cms or g cms ) with no luck

User is always mapped to "logcise":

http Protocol 'gsi'
http Name 'locsices'
http Host '[::ffff:10.10.150.1]'
http Vorg 'cms cms'
http Role 'NULL NULL'
http Grps '/cms /cms/escms'

Regards, I

**Comment by wyang007** [comment]
> Dear,
> I nee help to deploy http plugin for xrootd servers. The xrootd version we are runnin right now is xrootd-server-5.1.1-1.
> For sec.protocol we are using xrootd-voms-pluging and works fine.
> 
> But with the http/macaroons pluging we have some troubles, that is the config file:
> if exec xrootd
> xrd.protocol http:1094 /usr/lib64/libXrdHttp.so
> http.secxtractor libXrdHttpVOMS.so

There is a couple of things you need to straight out: xrootd-voms-pluging and libXrdHttpVoms.so are old packages working with xrootd 4. Since you are using xrotod 5.1.1, you should use
http.secxxtractor libXrdVoms.so
Similarly, in xroot protocol, you should also use libXrdVoms.so.

In addition, you should not use the following deprecated things
http.cert /etc/grid-security/xrd/xrdcert.pem
http.key /etc/grid-security/xrd/xrdkey.pem

You should config TLS instead:
# TLS
xrd.tls /etc/grid-security/xrd/xrdcert.pem /etc/grid-security/xrd/xrdkey.pem
xrd.tlsca certdir /etc/grid-security/certificates
xrootd.tls capable all

Ref: https://xrootd.slac.stanford.edu/doc/dev51/xrd_config.htm

I will reply to your auth_file issue next

**Comment by wyang007** [comment]
I don't know whether this is just a cut-n-paste issue. for example, "t writecmsdata ..." line is long. You must use a "\" at line end to continue. 

As Brian said, /cms/escms. not /escms

More specific lines in auth DB should come first. so in your case: 

g /cms/escms readcmsdata writecmsdata
g /cms/GGUSExpert readcmsdata writecmsdata
g /cms/TEAM readcmsdata writecmsdata
g /cms/ALARM readcmsdata writecmsdata
g /cms readcmsdata writecmsuser

(note /cms is the last line)

> [root@pool09 ~]# cat /etc/xrootd/Authfile
> t writecmsdata /store/backfill/ a 
> /store/data/ a 
> /store/generator/ a 
> /store/group/ a 
> /store/hidata/ a 
> /store/himc/ a 
> /store/mc/ a 
> /store/PhEDEx_LoadTest07/ a 
> /store/relval/ a 
> /store/results/ a 
> /store/temp/ a 
> /store/unmerged/ a 
> /store/user/ a
> 
> t readcmsdata /store/unmerged/ a 
> /store/ lr
> 
> t writecmsuser /store/temp/ a 
> /store/user/ a
> 
> # CMS users have full access to their own directory, and read for CMS
> # While xrootd allows the user to _attempt_ any operation - even in other user's
> # home directories - the underlying filesystem also has its internal permissions and will further
> # limit things.
> g /cms readcmsdata writecmsuser
> g /escms readcmsdata writecmsdata
> g /cms/GGUSExpert readcmsdata writecmsdata
> g /cms/TEAM readcmsdata writecmsdata
> g /cms/ALARM readcmsdata writecmsdata

**Comment by wyang007** [comment]
I guess github doesn't like back slash

**Comment by wyang007** [comment]
I mean you can continue a line in auth DB with a backslash

**Comment by cabrillo** [comment]
Hi Wei, 
   yes, there is a copy paste (at the end of the line there is a "\" ). i just modify the Authfile, in order to add at the end of the file the g /cms  group.

t readcmsdata  /store/unmerged/          a  "\"
                         /store/                           lr

t writecmsuser /store/temp/              a  "\"
                         /store/user/              a


g /cms/escms            readcmsdata writecmsuser
g /cms/GGUSExpert readcmsdata writecmsdata
g /cms/TEAM            readcmsdata writecmsdata
g /cms/ALARM          readcmsdata writecmsdata
g /cms                       readcmsdata writecmsuser

I just restart agents and test again, but the problem is still the same:

210601 12:01:32 7731 anon.0:33@login1 sysXrdHttp: Protocol matched. https: 1
210601 12:01:32 7731 sysXrdHttp:  Reset
210601 12:01:32 7731 sysXrdHttp:  XrdHttpReq request ended.
210601 12:01:32 7731 anon.0:33@login1 sysXrdHttp:  Setting host: [::ffff:10.10.150.1]
http Protocol 'gsi'
http Name 'locsices'
http Host '[::ffff:10.10.150.1]'
http Vorg 'cms cms'
http Role 'NULL NULL'
http Grps '/cms /cms/escms'
http Caps ''
http Pidn ''
http Crlen 0
http ueid  0
http uid   0
http gid   0
210601 12:01:32 7731 sysXrdHttp: read 203 of 1048576 bytes
210601 12:01:32 7731 XrootdBridge: locsices.3:33@login1 login as locsices
210601 12:01:32 7731 locsices.3:33@login1 sysXrdHttp: Process is exiting rc:1
210601 12:01:32 7731 acc_Audit: locsices.3:33@login1 deny gsi locsices@[::ffff:10.10.150.1] create /store/temp/stor.davix.01
210601 12:01:32 7731 ofs_open: locsices.3:33@login1 Unable to create /store/temp/stor.davix.01; permission denied

Regards, I

**Comment by abh3** [comment]
One quick correction

On Tue, 1 Jun 2021, Wei Yang wrote:
> You should config TLS instead:
> # TLS
> xrd.tls /etc/grid-security/xrd/xrdcert.pem /etc/grid-security/xrd/xrdkey.pem
> xrd.tlsca certdir /etc/grid-security/certificates
> xrootd.tls capable all
The last line should be

xrootd.tls capable all -data

Otherwise you will disable data optimization in the server and that is not 
what you want to do.

Andy

**Comment by wyang007** [comment]
Did you get a chance to change the voms extractor to libXrdVoms.so ?

**Comment by cabrillo** [comment]
Humm, 
I think so...but are the same, isn't it? 

[root@pool09 ~]# ls -la /usr/lib64/libXrdVoms-5.so 
-rwxr-xr-x 1 root root 49224 Mar 10 18:43 /usr/lib64/libXrdVoms-5.so
[root@pool09 ~]# ls -la /usr/lib64/libXrdHttpVOMS-5.so 
lrwxrwxrwx 1 root root 15 Apr 13 10:21 /usr/lib64/libXrdHttpVOMS-5.so -> libXrdVoms-5.so

well I just check it again, but the the output is the same.
regard, I

**Comment by wyang007** [comment]
In the config file, you can use /usr/lib64/libXrdHttpVOMS-5.so or you can use /usr/lib64/libXrdHttpVOMS.so. If you use the latter, xrootd 5 will automatically search with -5 appended.

**Comment by xrootd-dev** [comment]
Please NEVER specify the '-5' in the library name. It may or may not work 
and we will refuse tio fix if it doesn't. This is a big NO-NO.

**Comment by abh3** [comment]
We made the symlink to provide for backward compatability. All I can say 
is that the voms cert presented is mapping the name to what you see and if 
that is incorrect (or what you don't expect) them it's normal for 
authorization to fail. What mapping did you expect to have?

**Comment by cabrillo** [comment]
Dear,
  No, I not using the -5.so extension:
  if exec xrootd
  xrd.protocol http:1094 /usr/lib64/libXrdHttp.so
  http.secxtractor libXrdVoms.so
  #http.cadir /etc/grid-security/certificates/
  #http.cert /etc/grid-security/xrd/xrdcert.pem
  #http.key /etc/grid-security/xrd/xrdkey.pem

  #TLS config
  xrd.tls /etc/grid-security/xrd/xrdcert.pem /etc/grid-security/xrd/xrdkey.pem
  xrd.tlsca certdir /etc/grid-security/certificates
  xrootd.tls capable all -data
  http.exthandler xrdtpc libXrdHttpTPC.so
  http.exthandler xrdmacaroons libXrdMacaroons.so
  http.desthttps yes
  http.trace all -debug
  macaroons.secretkey /etc/xrootd/macaroon-secret
  #macaroons.onmissing allow
  macaroons.trace all
  ofs.authlib libXrdMacaroons.so
  http.listingdeny yes
  #http.staticpreload http://static/robots.txt /etc/xrootd/robots.txt  
  #http.header2cgi Authorization authz
fi

I hope to be maped to group cms (g /cms or /cms/escms) but not as **logsices**:
210601 21:01:19 7828 anon.0:33@login1 sysXrdHttp:  Setting host: [::ffff:10.10.150.1]
http Protocol 'gsi'
http Name 'locsices'
http Host '[::ffff:10.10.150.1]'
http Vorg 'cms cms'
http Role 'NULL NULL'
http Grps '/cms /cms/escms'
http Caps ''
http Pidn ''
http Crlen 0
http ueid  0
http uid   0
http gid   0
210601 21:01:19 7828 sysXrdHttp: read 204 of 1048576 bytes
210601 21:01:19 7828 XrootdBridge: locsices.3:33@login1 login as locsices
....
210601 21:01:19 7828 acc_Audit: locsices.3:33@login1 deny gsi locsices@[::ffff:10.10.150.1] create /store/temp/stor.davix.011
....
Regards, I

**Comment by abh3** [comment]
So, two things that I think are preventsing success:
1) There should be a "ofs.authorize" directive in the config. It can go next to the authlib directive (before or after). Without it no authorization take place.
2) The XrdMacaroons library is a stackable thing AFAIK and I really don't know why you need it if all if you just want to do normal authorization using the authfile.  However, in order to stack macaroons on top of the normal authorization scheme the authlib directive should be:
ofs.authlib ++ libXrdMacaroons.so
3) You said "not as locsices" so what were you expecting?

**Comment by cabrillo** [comment]
Dear;
I think i put the complet conf file ar the begining of the thread (https://github.com/xrootd/xrootd/files/6516114/xrootd-clustered.log)..xrootd server is working fine, and auth works too (using v5.1.1-1 in other 8 servers), this issue is only for http config and Third-Party-Copy

1.- ofs authorized is set to 1:
ofs.authorize 1
acc.authdb /etc/xrootd/Authfile
acc.audit deny grant

2.- I have been following this wiki (we are a CMS site), but this is only for v4 not v5 and use lcmaps instead of voms pluging:
     https://twiki.cern.ch/twiki/bin/view/CMSPublic/XRootDoverHTTP
    (This is working for me fine against the Authfile
     sec.protocol /usr/lib64 gsi -dlgpxy:1 -exppxy:creds -ca:1 -crl:3 -gridmap:/dev/null -cert:/etc/grid-security/xrd/xrdcert.pem -key:/etc/grid-security/xrd/xrdkey.pem -certdir:/etc/grid-security/certificates -vomsfun:/usr/lib64/libXrdVoms.so -vomsfunparms:certfmt=raw|vos=ops,dteam,cms|grps=/cms /escms/,/cms,/escms,/cms/GGUSExpert,/cms/TEAM,/cms/ALARM,/ops,/dteam|dbg

3. "locsices" or any other user should be fine if the http Grps '/cms /cms/escms' o  http Vorg 'cms cms'  was taken to match the authfile like xrootd is doing...or maybe i am wrong and http auth model is other?

regards, I

**Comment by abh3** [comment]
Yes, I used the last posting of the config file for my suggestions. BTW you can drop the '1' in ofs.authorize as the mere specification of the directive enables authlib. I am not familiar with the http 2rd party copy implementation when it comes to macaroons vs SciTokens. Presumably, they are interchangeable but it's not clear from the code whether or not they actually are. Other than using the '++' notation there are no differences. Unfortunately, when the macaroons plugin was written it didn't follow normal guidelines and used some internal interfaces which make it hardwired to a particular authorization scheme. Switching to '++' avoids that problem but the old code is still there so it should work without the '++', though I recommend moving to it for R5. How lcmaps interacts with the whole thing is also a mystery since it would be used on the xroot side but not necessarily the http side. Best to get advice from the relevant OSG people. Now, according to your trace, the proper group and vorg mapping is taking place so it should have worked other than the interposition of macaroons which may be mapping in a completely different way. Unfortunately, due to how the plugin was written there is no way to display the final result except via gdb.

**Comment by bbockelm** [comment]
Hi @cabrillo -

Looking over the thread, it's hard to determine what your current Authfile and configuration file are.

For example, the Authfile posting above is being processed as Markdown, meaning that relevant control characters are being removed (maybe - or perhaps they're missing).

Similarly, you mention using LCMAPS earlier in the thread but the configuration snippet you include doesn't actually show it.

Can you please try posting the latest attempt, either as attached files (as you did with the logs) or as code postings (using the ``` characters to delineate) so the config files aren't interpreted as control characters?

Thanks,

Brian

**Comment by bbockelm** [comment]
To answer a few other questions:

- When there's no mapfile present, XRootD defaults to parsing the DN to use as a name.  I assume you have `/CN= locsices` in your DN somewhere?
- The fact you're seeing a line like `210601 21:01:19 7828 acc_Audit: locsices.3:33@login1 deny gsi locsices@[::ffff:10.10.150.1] create /store/temp/stor.davix.011` and you're not using macaroons is good news - it means that you aren't hitting any issues from the Macaroons library in the current config and passing through to the default one.

I was able to reproduce a similar error message and log line when I had a slightly incorrect authfile and it was fixed when I fixed the authfile.  So, I suspect we should focus on that.
